cmd/scollector/collectors/procstats_linux.go

```diff
@@ -36,171 +28,165 @@ var CPU_FIELDS = []string{
+}
+
 func c_procstats_linux() (opentsdb.MultiDataPoint, error) {
 	var md opentsdb.MultiDataPoint
 	var Error error
-	if err := readLine("/proc/uptime", func(s string) error {
-		m := uptimeRE.FindStringSubmatch(s)
-		if m == nil {
-			return nil
-		}
-		Add(&md, "linux.uptime_total", m[1], nil, metadata.Gauge, metadata.Second, osSystemUptimeDesc)
-		Add(&md, "linux.uptime_now", m[2], nil, metadata.Gauge, metadata.Second, "")
-		Add(&md, osSystemUptime, m[1], nil, metadata.Gauge, metadata.Second, osSystemUptimeDesc)
-		return nil
-	}); err != nil {
-		Error = err
-	}
-	mem := make(map[string]float64)
+	var sys unix.Sysinfo_t
+
+	unix.Sysinfo(&sys)
+
+	Add(&md, "linux.uptime_total", sys.Uptime, nil, metadata.Gauge, metadata.Second, osSystemUptimeDesc)
+	Add(&md, osSystemUptime, sys.Uptime, nil, metadata.Gauge, metadata.Second, osSystemUptimeDesc)
 	if err := readLine("/proc/meminfo", func(s string) error {
-		m := meminfoRE.FindStringSubmatch(s)
+		s = strings.TrimSuffix(s, " kB")
+		m := strings.Split(s, ":")
 		if m == nil {
 			return nil
 		}
-		i, err := strconv.ParseFloat(m[2], 64)
-		if err != nil {
-			return err
-		}
-		mem[m[1]] = i
-		Add(&md, "linux.mem."+strings.ToLower(m[1]), m[2], nil, metadata.Gauge, metadata.KBytes, "")
+		m[1] = strings.TrimSpace(m[1])
+		Add(&md, "linux.mem."+strings.ToLower(m[0]), m[1], nil, metadata.Gauge, metadata.KBytes, "")
 		return nil
 	}); err != nil {
 		Error = err
+		slog.Errorln(err)
 	}
-	Add(&md, osMemTotal, int(mem["MemTotal"])*1024, nil, metadata.Gauge, metadata.Bytes, osMemTotalDesc)
-	Add(&md, osMemFree, (int(mem["MemFree"])+int(mem["Buffers"])+int(mem["Cached"]))*1024, nil, metadata.Gauge, metadata.Bytes, osMemFreeDesc)
-	Add(&md, osMemUsed, (int(mem["MemTotal"])-(int(mem["MemFree"])+int(mem["Buffers"])+int(mem["Cached"])))*1024, nil, metadata.Gauge, metadata.Bytes, osMemUsedDesc)
-	if mem["MemTotal"] != 0 {
-		Add(&md, osMemPctFree, (mem["MemFree"]+mem["Buffers"]+mem["Cached"])/mem["MemTotal"]*100, nil, metadata.Gauge, metadata.Pct, osMemFreeDesc)
+	Add(&md, osMemTotal, sys.Totalram*uint64(sys.Unit), nil, metadata.Gauge, metadata.Bytes, osMemTotalDesc)
+	Add(&md, osMemFree, sys.Freeram*uint64(sys.Unit), nil, metadata.Gauge, metadata.Bytes, osMemFreeDesc)
+	Add(&md, osMemUsed, (sys.Totalram-sys.Freeram)*uint64(sys.Unit), nil, metadata.Gauge, metadata.Bytes, osMemUsedDesc)
+	Add(&md, "linux.loadavg.1_min", sys.Loads[0], nil, metadata.Gauge, metadata.Load, "")
+	Add(&md, "linux.loadavg.5_min", sys.Loads[1], nil, metadata.Gauge, metadata.Load, "")
+	Add(&md, "linux.loadavg.15_min", sys.Loads[2], nil, metadata.Gauge, metadata.Load, "")
+	Add(&md, "linux.loadavg.total_threads", sys.Procs, nil, metadata.Gauge, metadata.Process, "")
+	if sys.Totalram != 0 {
+		Add(&md, osMemPctFree, (sys.Freeram)/(sys.Totalram)*100, nil, metadata.Gauge, metadata.Pct, osMemFreeDesc)
 	}
 	if err := readLine("/proc/vmstat", func(s string) error {
-		m := vmstatRE.FindStringSubmatch(s)
+		m := strings.Split(s, " ")
```

Comment:
It could be, but there is no need to check for tabs, multiple spaces since the separator is always a single space. If you look at the code for fields, it's a lot more complex and there's a lot more overhead. I have checked it in simple benchmark and for a single space Split is roughly 3 times faster than Fields. The longer the line the larger the margin.

The idea behind this patches is to get the checks faster (lighter) so I can increase their frequency with minimal load increase.